src/ray/gcs/subscription_executor.cc

```diff
@@ -72,43 +100,50 @@ template <typename ID, typename Data, typename Table>
 Status SubscriptionExecutor<ID, Data, Table>::AsyncSubscribe(
     const ClientID &client_id, const ID &id, const SubscribeCallback<ID, Data> &subscribe,
     const StatusCallback &done) {
-  Status status = AsyncSubscribe(client_id, nullptr, nullptr);
-  if (!status.ok()) {
-    return status;
-  }
-
-  auto on_done = [this, done, id](Status status) {
-    if (!status.ok()) {
-      std::lock_guard<std::mutex> lock(mutex_);
-      id_to_callback_map_.erase(id);
-    }
-    if (done != nullptr) {
-      done(status);
-    }
-  };
 
   {
-    std::lock_guard<std::mutex> lock(mutex_);
+    std::unique_lock<std::mutex> lock(mutex_);
     const auto it = id_to_callback_map_.find(id);
```

Comment:
The check here might not work. If there are two subscription request happens at the same time,  and both requests use the same ID, then both requests might receive ok status as return status,  but only one subscribe callback will registered. Because the protection domain of the lock is not big enough.

Reply:
The underlying problem is that the insertion for the map is moved to the callback, thus the check and insertion are no longer in one place. A fix can be to move the insertion outside of the callback to after the check.